### PR TITLE
Replace metricsmap-bpf-prom-sync with Prometheus Collector pattern

### LINKED ISSUE #27058: Replace Metricsmap-bpf-prom-sync with Prometheus Collector pattern
I was reviewing some work that @christarazi and @asauber were putting together reviewing our controllers/metrics and noticed that we seem to have a controller whose primarily responsibility is to periodically read values from a BPF map and update Prometheus metrics values based on the values in the map.

This approach has a few flaws, in particular:
- Metrics values are always slightly off. Prometheus scrapes periodically already, having another periodic timer means metrics can be weirdly skewed because of these two timers.
- Metrics values might stop getting updates if this controller stops running or doesn't startup successfully

From my previous experience and knowledge of other Go projects that need to deal with a similar problem, I believe there is a better alternative. The alternative is to implement the Prometheus Collector interface. https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics documents the use-case quite well, but in summary, the idea is that instead of managing counters, whenever Prometheus scrapes your application, it will call Collect() on your collector, and you then get the current value of all your metrics and send them on the channel given to you by Collect(). Basically, this eliminates the extra periodic metrics sync loop happening inside the controller, instead getting metrics values at scrape time. This is an approach taken by Prometheus node-exporter and cadvisor which read this information from /proc and /sys at scrape time.

This eliminates state, avoids metrics value accuracy, and overall is a simpler design that follows existing best practices for metrics collection.

### PR DESCRIPTION
**Pull request by carnerito**
Previously, the 'metricsmap-bpf-prom-sync' controller was responsible for periodically collecting cilium_datapath drop
and forward metrics.
This PR introduces the metricsmapCollector, which implements the Prometheus Collector interface within the
metricsmap package. As a result, the aforementioned controller has been removed.
The metricsmapCollector is registered within the global Prometheus registry during the initialization of metricsmap.
The metrics.Register function has been modified to propagate errors from the registry.Register function instead of
simply overriding it with nil.
The metricsmapCollector comprises two metrics maps: forwardedMetricsMap and droppedMetricsMap.
These maps are populated within a callback function passed to the IterateWithCallback function.
This approach serves two primary purposes:
1. Separation of Map Iteration and Metric Update: By separating the iteration over the BPF map and the updating of
    Prometheus metrics, the implementation ensures that no partial metrics are exposed in case of map iteration failure.
2. Normalization of exposed Metrics: Unlike the statement in the bpf/lib/metrics.h comments, which suggests exposing
    only one reason label for forwarded metrics, the eBPF code exposes multiple reasons. Through testing, it was found
    that reasons 0 and 3 were exposed which triggered this error in prometheus client:
    https://github.com/prometheus/client_golang/issues/242

Furthermore, cilium command has an indirect dependency on metricsmap package. To prevent metrics map initialization
from happening each time cilium command is executed, metricsmap is converted to hive Cell and injected in cilium daemon
Infrastructure module. For details see this comment: https://github.com/cilium/cilium/pull/27370#issuecomment-1710805663.

Fixes: #27058

Signed-off-by: Boris Petrovic <carnerito.b@gmail.com>

### CODE REVIEW

**Review by dylandreimerink — APPROVED** [review]

**Comment by dylandreimerink** [comment]
Hi @carnerito, thank you for this contribution. Could you sign off on your commit? https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#developer-s-certificate-of-origin

**Comment by carnerito** [comment]
@dylandreimerink Can I squash all commits and force-push?

File: Documentation/bpf/toolchain.rst
Comment (zacharysarah) [inline Documentation/bpf/toolchain.rst]: Who is "us"? I know this is inherited language, but it's an opportunity to be specific and helpful rather than remaining vague.

File: Documentation/community/governance/commit_access.rst
Comment (zacharysarah) [inline Documentation/community/governance/commit_access.rst]: Use present tense
```suggestion
`Code of Conduct`_ violations by community members are discussed and
```

File: Documentation/community/governance/commit_access.rst
Comment (zacharysarah) [inline Documentation/community/governance/commit_access.rst]: I think we can be gentler.
```suggestion
Votes are done in the Slack channel ``#committers`` using Slack polls.
Declining to vote is an implicit abstention.
```

File: Documentation/community/roadmap.rst
Comment (zacharysarah) [inline Documentation/community/roadmap.rst]: Avoid "we", use plain, clear language, specific is better than general, future tense is OK here
```suggestion
As the project grows in complexity, it is increasingly important
to be able to divide the project into more manageable chunks. To achieve this, contributors
are modularizing the codebase and going from a tightly coupled design
(one large initialization and configuration) to a more loosely coupled design
of mostly self-contained modules. This will make Cilium internals easier to
comprehend, test, and extend.
```

File: Documentation/community/roadmap.rst
Comment (zacharysarah) [inline Documentation/community/roadmap.rst]: Use plain, clear language (avoid "please")
```suggestion
or ideas, join us on the ``#sig-modularization`` channel on `Cilium
```

File: Documentation/configuration/argocd-issues.rst
Comment (zacharysarah) [inline Documentation/configuration/argocd-issues.rst]: Use clear, plain language
```suggestion
    When using ``helm template``, set
```

File: Documentation/configuration/argocd-issues.rst
Comment (zacharysarah) [inline Documentation/configuration/argocd-issues.rst]: accessibility nit
```suggestion
If you have an issue with Argo CD that's not outlined above, review this `list
```

File: Documentation/glossary.rst
Comment (zacharysarah) [inline Documentation/glossary.rst]: wording nit
```suggestion
on `Cilium Slack`_. Feel free to extend this document with terms you expected
```

File: Documentation/installation/cni-chaining-generic-veth.rst
Comment (zacharysarah) [inline Documentation/installation/cni-chaining-generic-veth.rst]: Who is "us"? I recognize this is inherited language, but it's a chance to be specific and helpful rather than continuing to be vague.

File: pkg/maps/metricsmap/metricsmap.go
Comment (zacharysarah) [inline pkg/maps/metricsmap/metricsmap.go]: Verb tense nit
```suggestion
	// IterateWithCallback iterates through BPF map for each reason and direction.
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@carnerito Thanks for the updates! Some minor stylistic changes are required, with strong encouragement to fix some of the inherited language touched by this PR.

File: pkg/maps/metricsmap/metricsmap.go
Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go]: @zacharysarah I'm really sorry for making confusion, but all changes you have proposed are not from my PR. I tried to amend the unsigned commit as suggested by @dylandreimerink and I've pushed someone else's commits to this branch by mistake.
The only thing that you have commented that is part of this PR is regarding the wording in metricsmap.go, and I have fixed that in [#d1988ae](https://github.com/cilium/cilium/pull/27370/commits/d1988ae903d5b7424af189c6bce16bf253c275d6)

File: Documentation/bpf/toolchain.rst
Comment (carnerito) [reply Documentation/bpf/toolchain.rst]: Not part of this PR.

File: Documentation/community/governance/commit_access.rst
Comment (carnerito) [reply Documentation/community/governance/commit_access.rst]: Not part of this PR.

File: Documentation/community/governance/commit_access.rst
Comment (carnerito) [reply Documentation/community/governance/commit_access.rst]: Not part of this PR.

File: Documentation/community/roadmap.rst
Comment (carnerito) [reply Documentation/community/roadmap.rst]: Not part of this PR.

File: Documentation/community/roadmap.rst
Comment (carnerito) [reply Documentation/community/roadmap.rst]: Not part of this PR.

File: Documentation/configuration/argocd-issues.rst
Comment (carnerito) [reply Documentation/configuration/argocd-issues.rst]: Not part of this PR.

File: Documentation/configuration/argocd-issues.rst
Comment (carnerito) [reply Documentation/configuration/argocd-issues.rst]: Not part of this PR.

File: Documentation/glossary.rst
Comment (carnerito) [reply Documentation/glossary.rst]: Not part of this PR.

File: Documentation/installation/cni-chaining-generic-veth.rst
Comment (carnerito) [reply Documentation/installation/cni-chaining-generic-veth.rst]: Not part of this PR.

**Review by christarazi — COMMENTED** [review]
Thanks for the PR! Could you rebase and squash all commits together?

**Comment by carnerito** [comment]
@christarazi Done.

File: pkg/maps/metricsmap/metricsmap.go
Comment (zacharysarah) [reply pkg/maps/metricsmap/metricsmap.go]: @carnerito
> all changes you have proposed are not from my PR

~~And yet your changes touch them, creating an opportunity to make docs better. Otherwise--when will docs ever improve?~~

~~I'm disappointed in your choice to refuse uncomplicated improvements, but I'll approve regardless.~~

UPDATE: See https://github.com/cilium/cilium/pull/27370#discussion_r1296458751 and following 👇🏻

**Review by zacharysarah — APPROVED** [review]
LGTM

File: pkg/maps/metricsmap/metricsmap.go
Comment (joestringer) [reply pkg/maps/metricsmap/metricsmap.go]: It looks to me like there was a misunderstanding here in the way that the changes got pushed into GitHub and the way that GitHub requested reviewers for these changes. From what I can tell, @carnerito never intended to make changes to these files in the PR, but the way that the PR was pushed into GitHub caused GitHub to pull @zacharysarah in for review.

In future it might be worth doing a quick double-check of the PR before marking it "ready for review" to make sure that GitHub presents the set of changes that you expected to present, so that GitHub only pulls in the codeowners for the areas that are being changed.

At this point it looks like @carnerito has been able to resolve this issue, maybe with a fresh push of the branch, and none of these comments are relevant to the changes in the PR any more. Either way it looks like this has been resolved in the end. @carnerito apologies for the confusion around this, thanks again for contributing these changes to Cilium :-)

Comment (zacharysarah) [reply pkg/maps/metricsmap/metricsmap.go]: @joestringer Thanks for the explanation. @carnerito Apologies for pushing back so strongly; I see the underlying issue now. Thanks for your contributions!

File: pkg/maps/metricsmap/metricsmap.go
Comment (christarazi) [inline pkg/maps/metricsmap/metricsmap.go:202]: Nit: unnecessary newline

**Review by christarazi — DISMISSED** [review]
Very cool PR! I was wondering whether we could make use of batch lookups instead of using `IterateWithCallback`, but realized that the metrics map doesn't have very many entries in it, so maybe it's just a nice-to-have thing. With the PR there's already a performance improvement because the controller used to run every 5s and now it's based on the scrape interval which is very likely to be >=15s.

**Comment by joestringer** [comment]
A couple of the CI jobs are failing due to timeouts ([example1](https://github.com/cilium/cilium/actions/runs/5884199682), [example2](https://github.com/cilium/cilium/actions/runs/5884200085) and there are many failures in the [main ginkgo test run](https://github.com/cilium/cilium/actions/runs/5884199995/job/15958487587). It seems like there's some kind of issue causing problems in the CI, but it's not immediately obvious what that issue is.

File: pkg/maps/metricsmap/metricsmap.go
Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go:202]: @christarazi Thanks for the review. I didn't know that maps have a BatchLookup method; I'll keep this in mind in the future. As for the newline, I was relying on gomt to do the formatting, but I completely agree that it would be better without it here.

**Comment by carnerito** [comment]
@joestringer This is something that happens from time to time, or do you think it is related to this PR?
Can we try re-running tests?

**Comment by joestringer** [comment]
I haven't seen this before, but I can re-run the CI tests to confirm.

**Comment by carnerito** [comment]
@joestringer Due to lack of experience with GitHub Actions and Cilium CI in general I have no idea how to proceed with this, do you have any suggestions?

It's really strange to me that same tests were passing before I pulled changes from main branch...

**Comment by carnerito** [comment]
@joestringer I've pulled the latest changes from the main branch and tests are not timing out anymore.

File: pkg/maps/metricsmap/metricsmap.go
Comment (derailed) [inline pkg/maps/metricsmap/metricsmap.go]: nit: no need for this tmp var. Call directly? `metrics.Register(newMetricsCollector())`

File: pkg/maps/metricsmap/metricsmap.go
Comment (derailed) [inline pkg/maps/metricsmap/metricsmap.go]: nit: superfluous else

**Review by derailed — DISMISSED** [review]
@carnerito Nice work Boris! Just a couple small picks...

**Comment by joestringer** [comment]
It's interesting to see that the same ginkgo failure is occurring for another completely different PR. I'm unsure about the cause. https://github.com/cilium/cilium/issues/27688

File: pkg/maps/metricsmap/metricsmap.go
Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go]: @derailed Thanks for review. I'll make changes after we resolve the issue with CI.

**Comment by carnerito** [comment]
> It's interesting to see that the same ginkgo failure is occurring for another completely different PR. I'm unsure about the cause. #27688

Indeed.

P.S. 
I'm AFK for the next 10 days so I won't be able to actively work on this until then. If you require any modifications to be made in this PR for testing a fix in CI, feel free to reach me out here or on Slack.

**Comment by joestringer** [comment]
We've identified https://github.com/cilium/cilium/issues/27688 as being triggered by other PRs also (and not reliably passing/failing), which suggests that the failure is unrelated to this PR. I think that the next step should be to rebase this PR against tip of main branch and only have the functional changes in the PR, then retrigger CI again. We'll need to pay close attention that CI doesn't point out any other issues with the actual PR, but if CI can pass with these changes then it should be fine to move forward.

**Comment by joestringer** [comment]
@carnerito I'll mark this PR as draft for now. When you get back, could you try to rebase afresh with the actual proposed changes for this PR, then click `Ready for review` at the bottom of the page? We can try afresh with the changes in the PR.

**Comment by carnerito** [comment]
> @carnerito I'll mark this PR as draft for now. When you get back, could you try to rebase afresh with the actual proposed changes for this PR, then click `Ready for review` at the bottom of the page? We can try afresh to test the intended changes in the PR.

I tried to revert to a commit with actual changes, but it didn't go well. As you can see here, this PR reverts some changes made after it was created. Pulling or rebasing onto main shows that everything is up to date, although `git diff metricsmap-collector upstream/main` reveals differences. Since I don't know how to fix this I've opened [#27740](https://github.com/cilium/cilium/pull/27740) with actual changes.

**Comment by carnerito** [comment]
@joestringer With a little help from @christarazi I've managed to fix the issues with this PR. It's ready for review.

**Comment by christarazi** [comment]
@carnerito What was the problem with the code such that the CI tests were failing?

**Comment by carnerito** [comment]
> @carnerito What was the problem with the code such that the CI tests were failing?

I didn't prove it is related to code from this PR. The same CI test failures are happening for other PRs too. @joestringer opened [#27688](https://github.com/cilium/cilium/issues/27688) to track this.

**Comment by christarazi** [comment]
I think that was for a specific failure for when the revert was applied, but this doesn't explain why the rest of CI especially the `ci-ginkgo` suite was timing out on almost all tests when the PR contained the original change. CI timing out was not the case when the revert was applied, so it does indeed seem like the original changes introduced in this PR are somehow causing test failures.

**Comment by carnerito** [comment]
I didn't understand [this comment](https://github.com/cilium/cilium/pull/27370#issuecomment-1693791258) than. Can we mark this PR as draft again until I investigate CI failures?

**Comment by christarazi** [comment]
Marked as draft. There may have been some confusion, but the issue identified (https://github.com/cilium/cilium/issues/27688) has nothing to do with the current failures (e.g. here's the same test case failing for different reasons: https://github.com/cilium/cilium/actions/runs/6004247752/job/16284443998#step:16:630) and is actually the same set of failures I investigated last week. It's a completely different error msg.

**Comment by carnerito** [comment]
@christarazi @joestringer I've figured out why Ginkgo tests were failing. It's a combination of multiple factors.

First:
- Ginkgo tests were failing because the kube-dns deployment was not configured on time -> [View Details](https://github.com/cilium/cilium/actions/runs/6004247752/job/16284443998#step:16:261)
- The kube-dns deployment was not configured on time since the [kube-dns service plumbing check](https://github.com/cilium/cilium/blob/5b0cf2cc3d6abcdf7b1a662691c9620237b31b51/test/helpers/kubectl.go#L2058) was hitting a 10s timeout, as seen in the logs.
- The plumbing check was timing out because the [Cilium commands that are run](https://github.com/cilium/cilium/blob/5b0cf2cc3d6abcdf7b1a662691c9620237b31b51/test/helpers/kubectl.go#L1862-L1884) to fetch Cilium state took more time than expected to finish.
- The delay mentioned in the previous bullet point is introduced by the metricsmap package initialization, which adds approximately 1s latency to any Cilium command (e.g., `cilium service list`) (I'm not sure why metricsmap is initialized since it is not invoked directly by any cilium command, so I guess it's indirect dependency).
- [#e2839ab](https://github.com/cilium/cilium/pull/27370/commits/e2839abbef9875419904e0ff1e535a9e0ee27993) fixes this issue.


Second:
I believe there is a bug in the CiliumExecContext function. [It repeats the same command 5 times without checking if the command has succeeded](https://github.com/cilium/cilium/blob/5b0cf2cc3d6abcdf7b1a662691c9620237b31b51/test/helpers/kubectl.go#L2911-L2920).
If we combine this behavior with the 1-second delay mentioned above and multiply that by the 2 kube-dns pods, we have a reason for exceeding the 10s limit when checking if kube-dns is validating service plumbing.

File: pkg/maps/metricsmap/metricsmap.go
Comment (christarazi) [inline pkg/maps/metricsmap/metricsmap.go]: Hm, the fact that this initialization occurs when a Cilium command is run first seems undesirable. Ideally, the daemon initializes all the metrics it needs upfront. The Cilium commands (which are separate binaries from the daemon) should not have to run any code to initialize metrics. 

I think what we could do here is to export `newMetricsMapCollector()` and have to be called within the `pkg/metrics/cell.go` code which is imported by the Agent Cell in `daemon/cmd/cells.go`.

cc @dylandreimerink, does this sound reasonable?

**Review by christarazi — COMMENTED** [review]
Thanks for the thorough investigation! 

Below is a comment on the metrics initialization code and regarding the bug in Cilium test helper functions, a separate PR would be good for that.

**Comment by carnerito** [comment]
PR that fixes bug in Cilium test helper function [#28131](https://github.com/cilium/cilium/pull/28131).

File: pkg/maps/metricsmap/metricsmap.go
Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go]: @dylandreimerink @christarazi Can you take a look [at latest commit](https://github.com/cilium/cilium/pull/27370/commits/54b1945191db724ac094c79e9785cf9f1b017d39). I have converted metricsmap to cell.Module which is invoked in `daemon/cmd/cells.go`. Now, only daemon will initialize metricsmap. I've tried to add it to `pkg/metrics/cell.go` as suggested, but than we have circular dependency issue
```
package github.com/cilium/cilium/api/v1/server
        imports github.com/cilium/cilium/pkg/metrics
        imports github.com/cilium/cilium/pkg/maps/metricsmap
        imports github.com/cilium/cilium/pkg/ebpf
        imports github.com/cilium/cilium/pkg/bpf
        imports github.com/cilium/cilium/pkg/controller
        imports github.com/cilium/cilium/pkg/metrics: import cycle not allowed
```
Furthermore, I've tried to add metricsmap collector to metricsmap module, like in Hive examples, with `cell.Metrics(newMetricsmapCollector)`, but I couldn't make it work that way. Even though all metricsmap metrics were registered, metricsmap Collector was not, so Collect method wasn't invoked on the scrape. I know that these metrics were registered because I've spawn separate goroutine that was incrementing counters periodically and I was able to scrape them. 
Is it possible to create custom Collector using metric types defined in `pkg/metrics/metric` package?

Comment (christarazi) [reply pkg/maps/metricsmap/metricsmap.go]: The structure looks good. I think it's unfortunate that there's an import cycle, which suggests that further refactoring is needed, but let's not tackle that in this PR. It would be good to keep that as a note in a code comment on the metricsmap cell. 

I'll let Dylan get to the finer details here.

**Comment by lmb** [comment]
@christarazi you set the `dont-merge/preview-only` label. Is that still appropriate?

**Review by christarazi — COMMENTED** [review]
I think you can squash all three commits into one.

**Comment by carnerito** [comment]
Done.

**Comment by lmb** [comment]
Please resolve all conversations (including "outdated" ones) by clicking "Resolve conversation" if the feedback has been addressed. Otherwise merging is not possible.

**Comment by carnerito** [comment]
@lmb Done. All conversations are resolved

**Comment by lmb** [comment]
@carnerito you've squashed the commits but forgot to adjust the commit message.

**Comment by carnerito** [comment]
> @carnerito you've squashed the commits but forgot to adjust the commit message.

@lmb Sorry for that, it's fixed now.

**Comment by christarazi** [comment]
@carnerito There seems to be a legitimate failure in the Ginkgo tests regarding BPF metrics:

```
• Failure [106.658 seconds]
K8sDatapathServicesTest
/home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:461
  Checks N/S loadbalancing
  /home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:461
    Supports IPv4 fragments [It]
    /home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:515

    Failed to account for INGRESS IPv4 fragments to 10.96.194.76 in BPF metrics
    Expected
        <[]int | len:2, cap:2>: [0, 0]
    To satisfy at least one of these matchers: [%!s(*matchers.EqualMatcher=&{[0 4]}) %!s(*matchers.EqualMatcher=&{[4 0]})]
```

https://github.com/cilium/cilium/actions/runs/6378208180/job/17308444160

**Comment by carnerito** [comment]
@christarazi Can you rerun tests please. The `cilium bpf metrics` commands are currently reliant on the metrics map. I've made a tweak by moving the metrics map initialization to the var block. However, do note that the Prometheus collector is registered only when the metrics map Cell is invoked. I admit it's a bit clunky, but after some digging it seemed to be the best workaround.

**Comment by christarazi** [comment]
@carnerito The easiest way is to run the ginkgo test suite by passing the current kubeconfig that points to a local kind cluster.

File: pkg/maps/metricsmap/metricsmap.go
Comment (derailed) [inline pkg/maps/metricsmap/metricsmap.go:230]: We could perhaps dry this up a bit by creating a metricsmap generic type and hang methods to upsert or even exporting metrics on the map vs using functions? i.e
```go
type metricsMap[k labels] map[k]metricValues

func (m metricsMap[labels]) upsert(l labels, b, c float64) {
	m[l] = m[l].inc(b, c)
}
// Then...
err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
		if key.IsDrop() {
		  m.droppedMetricMap.upsert(dropLabels{...}, values.Bytes, value.Counts)
		} else {
		  m.forwardedMetricMap.upsert(forwardLabels{...}, values.Bytes, value.Counts)
		}
                ...
}
```

**Review by derailed — COMMENTED** [review]
@carnerito Thank you Boris for this PR. Nice work!!

File: pkg/maps/metricsmap/metricsmap.go
Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go:230]: @derailed Thanks for review! 
I've created a generic `promMetrics` map with upsert() method as suggested. Code in a callback function looks bit cleaner now. 
Regarding exporting, I agree that it would be better if we could have `m.droppedMetricMap.export` and `m.forwardedMetricMap.export` instead those two for loops at the end of callback function, but in that case we would have to associate prometheus.Desc with labels and values. Since we have two values under same key (labels) in metrics map I don't see any elegant way to do this without rewriting most of the logic, if you have any hints, I'm willing to try :)

Comment (carnerito) [reply pkg/maps/metricsmap/metricsmap.go:230]: @derailed Hi, do you have any updates on this?

**Comment by carnerito** [comment]
@aanm Cilium build was failing because of [unused import](https://github.com/cilium/cilium/actions/runs/6704188693/job/18216131655?pr=27370#step:21:396). It's fixed now, could you please rerun tests?

**Comment by aanm** [comment]
@carnerito Thanks, can you add the last commit as part of https://github.com/cilium/cilium/pull/27370/commits/8e741dfcb92ab47b666450c957b4a58d61b1713a? We try that each commit is in a "buildable state"

**Comment by carnerito** [comment]
@aanm Just to be sure, do you want me to squash all commits and force-push?

**Comment by aanm** [comment]
> @aanm Just to be sure, do you want me to squash all commits and force-push?

@carnerito I didn't look at all commits but now that you mentioned it, yes please squash all commits together and force-push. Thank you!

**Comment by carnerito** [comment]
@aanm Done

File: pkg/maps/metricsmap/metricsmap.go
Comment (derailed) [reply pkg/maps/metricsmap/metricsmap.go:230]: @carnerito I think what you have now is indeed fine.

**Review by derailed — APPROVED** [review]
@carnerito Thank you for this PR and updates Boris!!

**Comment by joestringer** [comment]
Required tests are passing. Seems to be mainly waiting on @cilium/sig-foundations review due to the cells.go change.

**Comment by joestringer** [comment]
Removing my review, I was only involved here as triager trying to keep PRs moving.

**Review by christarazi — APPROVED** [review]

**Review by dylandreimerink — APPROVED** [review]

**Comment by pchaigno** [comment]
Is it me or this pull request involuntary renamed a bunch of metrics? For example, what used to be called `cilium_drop_count_total` is now called `cilium_datapath_drop_count_total`. Isn't that a breaking change?

**Comment by carnerito** [comment]
@pchaigno Yes, it's fixed in #29226

**Comment by pchaigno** [comment]
Ah, perfect! Thanks! The fix simply didn't reach my PR yet :slightly_smiling_face:

**Pull request merged**
